Content.Client/Pinpointer/UI/NavMapControl.cs

```diff
@@ -325,38 +364,84 @@ protected override void Draw(DrawingHandleScreen handle)
+                    var position = _transformSystem.GetInvWorldMatrix(xform).Transform(mapPos.Position) - offset;
                     position = Scale(new Vector2(position.X, -position.Y));
 
-                    handle.DrawCircle(position, MinimapScale / 2f, value.Color);
+                    float f = 2.5f;
+                    var rect = new UIBox2(position.X - float.Sqrt(MinimapScale) * f, position.Y - float.Sqrt(MinimapScale) * f, position.X + float.Sqrt(MinimapScale) * f, position.Y + float.Sqrt(MinimapScale) * f);
+
+                    if (value.Texture != null)
+                        handle.DrawTextureRect(value.Texture, rect, value.Color);
+                    else
+                        handle.DrawCircle(position, float.Sqrt(MinimapScale) * 2f, value.Color);
                 }
             }
         }
 
         // Beacons
-        var labelOffset = new Vector2(0.5f, 0.5f) * MinimapScale;
-        var rectBuffer = new Vector2(5f, 3f);
+        if (_beacons.Pressed)
+        {
+            var labelOffset = new Vector2(0.5f, 0.5f) * MinimapScale;
+            var rectBuffer = new Vector2(5f, 3f);
+            var beaconColor = Color.FromSrgb(TileColor.WithAlpha(0.8f));
+
+            foreach (var beacon in navMap.Beacons)
+            {
+                var position = beacon.Position - offset;
+
+                position = Scale(position with { Y = -position.Y });
+
+                handle.DrawCircle(position, MinimapScale / 2f, beacon.Color);
+                var textDimensions = handle.GetDimensions(_font, beacon.Text, 1f);
+
+                var labelPosition = position + labelOffset;
+                handle.DrawRect(new UIBox2(labelPosition, labelPosition + textDimensions + rectBuffer * 2), beaconColor);
+                handle.DrawString(_font, labelPosition + rectBuffer, beacon.Text, beacon.Color);
+            }
+        }
+    }
+
+    protected override void FrameUpdate(FrameEventArgs args)
+    {
+        _updateTimer += args.DeltaSeconds;
 
-        foreach (var beacon in navMap.Beacons)
+        if (_updateTimer >= UpdateTime)
         {
-            var position = beacon.Position - offset;
+            _updateTimer -= UpdateTime;
 
-            position = Scale(position with { Y = -position.Y });
+            if (!_entManager.TryGetComponent<NavMapComponent>(MapUid, out var navMap))
+                return;
 
-            handle.DrawCircle(position, MinimapScale / 2f, beacon.Color);
-            var textDimensions = handle.GetDimensions(_font, beacon.Text, 1f);
+            if (!_entManager.TryGetComponent<MapGridComponent>(MapUid, out var grid))
+                return;
 
-            var labelPosition = position + labelOffset;
-            handle.DrawRect(new UIBox2(labelPosition, labelPosition + textDimensions + rectBuffer * 2), BeaconColor);
-            handle.DrawString(_font, labelPosition + rectBuffer, beacon.Text, beacon.Color);
+            TileGrid = GetDecodedTileChunks(navMap.Chunks, grid);
```

Comment:
This shouldn't be getting done every frame

Reply:
I've updated the code so that all the TryGetComponents (including those under the Draw method) only get called when the map UID is updated (i.e., only when the console UI is opened)